### PR TITLE
Taxonomy "double dropdown" editing interface on Learning Resources

### PR DESCRIPTION
**Pull request by jamiefolsom**
Hi @carsongee,

Thanks for offering to have a look at this PR. Couple of items to note.
- I have not written any tests yet;
- My commits not squashed;
- Otherwise, seems to work for the UX as I understand it.

Happy for your thoughts on style, terseness, functionality, or anything else.

Jamie

### CODE REVIEW

**Comment by pdpinch** [comment]
Not sure what the etiquette is here, but since you're looking for feedback, I added the "Needs Review" label.

But I also added the "work in progress" label too.

File: ui/static/ui/js/learning_resources.jsx [hunk 1]
```diff
@@ -6,73 +6,160 @@ define('learning_resources', [
...
+              terms: tuple.terms,
+              selectedTerms: selectedTerms
+            };
+            return newTuple;
+          } else {
+            return tuple;
+          }
+        });
+
+      thiz.setState({
```
Comment (carsongee) [inline ui/static/ui/js/learning_resources.jsx]: I don't think you should need to use `thiz` for these unless you need to access the parent `this` from within a closure.

Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx]: Good point. Fix via new commit just pushed.

**Comment by amir-qayyum-khan** [comment]
@jamiefolsom I added 4 Vocabularies

screen shot 2015-08-12 at 5 27 36 pm
 
##### But it is showing me only 2 vocabularies in learning resource panel also i am not able to search other Vocabularies

screen shot 2015-08-12 at 5 27 44 pm

screen shot 2015-08-12 at 5 34 06 pm

**Comment by pdpinch** [comment]
FYI, vocabulary facets don't show up in the listing view unless they are part of the result set. 

In other words, you need to add a vocabulary term to at least one learning resource before the vocabulary will appear in the left nav.

**Comment by amir-qayyum-khan** [comment]
@pdpinch  here i am not referring to left nav, I am talking about learning resource panel.  it is not showing all vocabularies that i added in taxonomy panel.

**Comment by amir-qayyum-khan** [comment]
@jamiefolsom 

In **Learning resource panel**, when i selects **second vocabulary** it shows me list of **terms** of first and second  **vocabularies**. 
- Should it only show list of **terms** related to that vocabulary?
- When I selects 3rd or nth vocabulary then it show me **terms** of  first and that selected **vocabulary**.  Is this right behaviour?

screen shot 2015-08-12 at 5 50 07 pm

screen shot 2015-08-12 at 5 50 25 pm

**Comment by amir-qayyum-khan** [comment]
@jamiefolsom 
When I selected **colour** as **vocabulary** and  choosed  **terms** of **vocabulary name** in second field and press **save**. It has not given me any error.

screen shot 2015-08-12 at 5 59 05 pm

screen shot 2015-08-12 at 5 59 17 pm

**Comment by amir-qayyum-khan** [comment]
@jamiefolsom 
When I selected **colour** as **vocabulary** and  choosed  **terms** of **vocabulary "name"** in second field then saved and closed the panel. 
When i reopened it terms are allocated to  **vocabulary "name"** instead of colour. 

##### This seem good but it should inform user if he is mapping wrong terms with vocabulary or it should only show relevant terms against a vocabulary.

screen shot 2015-08-12 at 5 59 05 pm

screen shot 2015-08-12 at 6 03 37 pm

**Comment by jamiefolsom** [comment]
Hi @amir-qayyum-khan, thanks for taking a look! You should only see terms in the term menu for the specific vocabulary you selected from the vocabulary menu. Sounds like a bug; I will look into it.

File: ui/static/ui/js/learning_resources.jsx [hunk 2]
```diff
@@ -6,73 +6,156 @@ define('learning_resources', [
...
+          <strong>{this.props.label}: </strong> {this.props.terms}
+        </li>
+      );
+    }
+  });
+
+  var VocabSelect2 = React.createClass({
+
+    render: function () {
+
```
Comment (amir-qayyum-khan) [inline ui/static/ui/js/learning_resources.jsx]: Please remove blank line after render?

File: ui/static/ui/js/learning_resources.jsx [hunk 3]
```diff
@@ -6,73 +6,156 @@ define('learning_resources', [
...
+      return (
+        <li className="applied-term list-group-item">
+          <strong>{this.props.label}: </strong> {this.props.terms}
+        </li>
+      );
+    }
+  });
+
+  var VocabSelect2 = React.createClass({
+
```
Comment (amir-qayyum-khan) [inline ui/static/ui/js/learning_resources.jsx]: Please remove blank line before render?

File: ui/static/ui/js/learning_resources.jsx [hunk 4]
```diff
@@ -6,73 +6,156 @@ define('learning_resources', [
...
-      var options = _.map(this.props.terms, function (term) {
+      return (
+        <li className="applied-term list-group-item">
+          <strong>{this.props.label}: </strong> {this.props.terms}
+        </li>
+      );
+    }
+  });
+
+  var VocabSelect2 = React.createClass({
```
Comment (amir-qayyum-khan) [inline ui/static/ui/js/learning_resources.jsx]: Why there is 2 with selector name? Should it be more clear like VocabSelect, VocabSelectForLR or any other?

File: ui/static/ui/js/learning_resources.jsx [hunk 5]
```diff
@@ -6,73 +6,156 @@ define('learning_resources', [
...
+    handleChange: function(e) {
+      var selectedValue = _.pluck(
+        _.filter(e.target.options, function(option) {
+          return option.selected && option.value !== null;
+        }), 'value');
+      this.props.updateVocabulary(_.pluck(this.props.vocabs, 'vocabulary'), selectedValue[0]);
+    }
+  });
+
+  var TermSelect2 = React.createClass({
```
Comment (amir-qayyum-khan) [inline ui/static/ui/js/learning_resources.jsx]: Why 2 is append with TermSelect because i havent seen any other TermSelect in the same file. Should it be more clear like TermSelect or TermSelectForLR or LRTermSelect or any other

File: ui/static/ui/js/learning_resources.jsx [hunk 6]
```diff
@@ -6,73 +6,156 @@ define('learning_resources', [
...
+
+    render: function () {
+
```
Comment (amir-qayyum-khan) [inline ui/static/ui/js/learning_resources.jsx]: please remove this line

File: ui/static/ui/js/learning_resources.jsx (hunk 5 above)
Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx]: The "2"s refer to the fact that the components are built with the Select2 component, in utils. But you're right -- that's unnecessary. I'll change those names.

File: ui/static/ui/js/learning_resources.jsx (hunk 3 above)
Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx]: Will do.

File: ui/static/ui/js/learning_resources.jsx (hunk 2 above)
Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx]: Yep -- there were comments in there, now gone -- will tighten up the whitespace. Thx.

File: ui/static/ui/js/learning_resources.jsx (hunk 6 above)
Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx]: Line 52?

**Comment by jamiefolsom** [comment]
Hi, @amir-qayyum-khan thanks for catching that issue (terms from other vocabularies getting duplicated). See [latest commit](https://github.com/mitodl/lore/commit/8ea6cc6e74c6b50e08355add33d82ccc197ae768) for the solution; I was not setting a `key` prop on the dynamically updating child component, TermSelect; without that, React didn't know how to update it properly. cc @pdpinch and @carsongee.

**Comment by carsongee** [comment]
Cool, thanks for tagging me in with the answer.  I didn't know about the key prop

**Comment by pdpinch** [comment]
Is that `key` prop a generic react requirement, or something specific to the Select2 implementation?

**Comment by ShawnMilo** [comment]
`key` is a generic React thing.

**Comment by jamiefolsom** [comment]
Yes, it's generic, but as I understand it, it's not required in all cases, only those in which the component has dynamic content, updated by a parent.

The fact that it's required in that case seems sort of odd, since (at least from my pt. of view), the expected behavior would be for an updated prop to result in a updated UI.

But I think it's because React is trying to be efficient in terms of how much of the DOM to change, so it reuses existing DOM nodes rather than replacing/deleting them, which in turn means it needs unique identifiers. Is that right, @ShawnMilo?

**Comment by ShawnMilo** [comment]
@jamiefolsom I don't know all the details. I just remember seeing it mentioned in one of the videos and that it had to be unique. I assumed from context it's exactly what you said (used to determine what parts of the DOM need updating), but I don't know of situations where it's optional.

I just scanned the video again and the brief explanation given is that you need it any time you have an array, so React can identify the same elements in an updated version of the array.

**Comment by jamiefolsom** [comment]
Makes sense. Thanks, @ShawnMilo.

File: ui/static/ui/js/utils.jsx [hunk 7]
```diff
@@ -109,8 +109,13 @@ define("utils", ["jquery", "lodash", "reactaddons", "select2"],
       //this function can be used only if the component has been mounted
       var thiz = this;
       var isMultiTerms = this.props.multiple;
+
       //not allowing the clear for the multi terms dropdown
-      var allowClear = !isMultiTerms;
+      var allowClear;
+      if (this.props.allowClear === undefined) {
```
Comment (noisecapella) [inline ui/static/ui/js/utils.jsx:115]: It seems like this needs an `else` block to assign allowClear

File: ui/static/ui/css/mit-lore.css [hunk 8]
```diff
@@ -498,6 +498,32 @@ ul.with-utility-features {
     word-wrap: break-word;
 }
 
+div#term-list-container .panel-heading {
```
Comment (noisecapella) [inline ui/static/ui/css/mit-lore.css:500]: Can you change the indenting to match the rest of the CSS?

File: ui/static/ui/js/utils.jsx (hunk 7 above)
Comment (jamiefolsom) [inline ui/static/ui/js/utils.jsx:115]: Good catch.

File: ui/static/ui/css/mit-lore.css (hunk 8 above)
Comment (jamiefolsom) [inline ui/static/ui/css/mit-lore.css:500]: Sure thing, will do.

File: ui/static/ui/js/utils.jsx (hunk 7 above)
Comment (jamiefolsom) [inline ui/static/ui/js/utils.jsx:115]: So correct me if I am wrong, but as written, this conditional essentially sets a default, if a value is _not_ passed into the component when it's created, so an else here would be redundant.

Comment (noisecapella) [inline ui/static/ui/js/utils.jsx:115]: `this.props.allowClear` is only being tested against `undefined`. What if the caller passes `allowClear={true}`?

File: ui/static/ui/js/learning_resources.jsx [hunk 9]
```diff
@@ -6,73 +6,179 @@ define('learning_resources', [
...
+        if (selectedTerms.length) {
+          return (
+            <TermListItem
+              key={vocab.vocabulary.id}
+              label={vocabularyName}
+              terms={selectedTerms.join(", ")}
+            />
+          );
+        }
+      }, this);
```
Comment (noisecapella) [inline ui/static/ui/js/learning_resources.jsx]: It doesn't seem like this argument is necessary here since there is no `this` within the closure

Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx]: Which one?

Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx]: ah, `this`?

File: ui/static/ui/js/utils.jsx (hunk 7 above)
Comment (jamiefolsom) [inline ui/static/ui/js/utils.jsx:115]: Right. Didn't notice the later code relying on allowClear, rather than this.props.allowClear. Thanks!

File: ui/static/ui/css/mit-lore.css (hunk 8 above)
Comment (jamiefolsom) [inline ui/static/ui/css/mit-lore.css:500]: Changed tabs to four spaces -- let me know if that fixed what you were seeing. Thanks.

**Comment by jamiefolsom** [comment]
Hey @pdpinch, I've addressed @noisecapella's suggestions. Others are of course welcome. I'm removing the WIP tag, since this is under review. It's rebased on master; only other outstanding step that I know of is to squash commits.

**Comment by jamiefolsom** [comment]
@bdero if you have a sec, it appears that a Travis build is still failing on this PR. I haven't rerun, or pushed a new commit. See https://travis-ci.org/mitodl/lore/builds/76001652#L866. It appears unrelated to my PR, but any thoughts you may have to offer would be welcome. Thanks!

**Comment by pdpinch** [comment]
You failed a test, which appears to be unrelated & due to a publicsuffixes.org being down:

> it appears that https://publicsuffix.org/list/effective_tld_names.dat has either changed or was offline, it appears up now, retrying

I'm rerunning Travis now.

**Comment by jamiefolsom** [comment]
@pdpinch Thanks!

**Comment by noisecapella** [comment]
I still need to look over this, I haven't made it completely through yet

File: ui/static/ui/js/utils.jsx [hunk 10]
```diff
@@ -109,8 +109,13 @@ define("utils", ["jquery", "lodash", "react", "select2"],
       //this function can be used only if the component has been mounted
       var thiz = this;
       var isMultiTerms = this.props.multiple;
+
       //not allowing the clear for the multi terms dropdown
```
Comment (noisecapella) [inline ui/static/ui/js/utils.jsx]: Can you update the comment according to the change you made?

Comment (jamiefolsom) [inline ui/static/ui/js/utils.jsx]: Done.

**Comment by noisecapella** [comment]
When I run the tests I get this warning:

```
Warning: flattenChildren(...): Encountered two children with the same key, `.$1`. Child keys must be unique; when two children share a key, only the first child will be used.
```

I narrowed it down to 'Assert that TermList renders properly'. Can you investigate? Duplicate keys are bad since React is effectively throwing one element away.

**Comment by noisecapella** [comment]
I've gotten the key error before in testing, it's usually just something to do with the test data being mocked.

**Comment by jamiefolsom** [comment]
Looks like the mock data in the test has two vocabularies with the same value for id. See: https://github.com/mitodl/lore/blob/feature/jf/taxonomy_double_dropdown/ui/jstests/test-learning-resource.jsx#L52-L73

I am going to change that, and see if it fixes this.

File: ui/static/ui/js/learning_resources.jsx [hunk 11]
```diff
@@ -6,73 +6,179 @@ define('learning_resources', [
...
+        <li className="applied-term list-group-item">
+          <strong>{this.props.label}: </strong> {this.props.terms}
+        </li>
+      );
+    }
+  });
+
+  var VocabSelect = React.createClass({
+    render: function () {
+      var options = [];
```
Comment (noisecapella) [inline ui/static/ui/js/learning_resources.jsx]: Can you combine this with the next line?

File: ui/static/ui/css/mit-lore.css (hunk 8 above)
Comment (noisecapella) [inline ui/static/ui/css/mit-lore.css:500]: Yes that's fine

File: ui/static/ui/js/learning_resources.jsx (hunk 11 above)
Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx]: I broke them out because I was getting line length complaints from the linter.

File: ui/jstests/test-learning-resource.jsx [hunk 12]
```diff
@@ -119,76 +128,117 @@ define(['QUnit', 'jquery', 'react', 'lodash', 'learning_resources',
...
-          assert.equal($terms1Select[0].selected, false);
-          assert.equal($terms1Select[0].selected, false);
+          var $allSelects = $node.find("#vocabularies select");
+          assert.equal($allSelects.size(), 2);
+
+          var $vocabSelect = $node.find($allSelects).first();
+          assert.equal($vocabSelect.size(), 1);
+
+          // first vocab, two options
+          var $vocabOptions = $($vocabSelect).find("option");
```
Comment (noisecapella) [inline ui/jstests/test-learning-resource.jsx]: If `$vocabSelect` is a jquery object you shouldn't need to use `$()` here

File: ui/jstests/test-learning-resource.jsx [hunk 13]
```diff
@@ -119,76 +128,117 @@ define(['QUnit', 'jquery', 'react', 'lodash', 'learning_resources',
...
+          assert.equal($vocabOptions.size(), 2);
+
+          assert.equal($vocabOptions[0].selected, true);
+          assert.equal($vocabOptions[1].selected, false);
 
           // TestUtils.Simulate.change only simulates a change event,
           // we need to update the value first ourselves
-          $($vocabSelect[0]).val("hard").trigger('change');
-          React.addons.TestUtils.Simulate.change($vocabSelect[0]);
+          $($vocabSelect).val("prerequisite").trigger('change');
```
Comment (noisecapella) [inline ui/jstests/test-learning-resource.jsx]: Same comment about `$()` here

File: ui/static/ui/js/learning_resources.jsx (hunk 11 above)
Comment (noisecapella) [inline ui/static/ui/js/learning_resources.jsx]: Would it be possible to move the function argument to the next line instead? Or just use `var options;`

File: ui/static/ui/js/learning_resources.jsx [hunk 14]
```diff
@@ -6,73 +6,179 @@ define('learning_resources', [
...
-        <label className="col-sm-6 control-label">
-        {this.props.vocabulary.name}
-          </label>
-        <div className="col-sm-6">
+      var currentVocabulary = {};
+      var selectedVocabulary = this.props.selectedVocabulary;
+
+      _.forEach(this.props.vocabs, function(vocab) {
+        if (vocab.vocabulary.slug === selectedVocabulary.slug) {
+          currentVocabulary = vocab;
```
Comment (noisecapella) [inline ui/static/ui/js/learning_resources.jsx:99]: Can you return false after this line? This will terminate the foreach early

File: ui/static/ui/js/learning_resources.jsx [hunk 15]
```diff
@@ -188,8 +299,10 @@ define('learning_resources', [
             });
 
           thiz.setState({
-            vocabulariesAndTerms: vocabulariesAndTerms
+            vocabulariesAndTerms: vocabulariesAndTerms,
+            selectedVocabulary: vocabulariesAndTerms[0].vocabulary
```
Comment (noisecapella) [inline ui/static/ui/js/learning_resources.jsx]: What happens if there are no vocabularies or terms for a resource?

File: ui/jstests/test-learning-resource.jsx [hunk 16]
```diff
@@ -119,76 +128,117 @@ define(['QUnit', 'jquery', 'react', 'lodash', 'learning_resources',
...
-            assert.equal($terms2Select[0].selected, true);
-            assert.equal($terms2Select[1].selected, false);
-            //the second vocabulary can be a multi select
-            $($vocabSelect[1])
-              .val(["notrequired", "required"])
+            assert.equal($vocabOptions[0].selected, false);
+            assert.equal($vocabOptions[1].selected, true);
+            // on selection of the second vocabulary, make sure the term selector updates.
+            var $termsSelect = $allSelects[1];
+            var $termsOptions = $($termsSelect).find("option");
```
Comment (noisecapella) [inline ui/jstests/test-learning-resource.jsx]: Remove `$()` here

File: ui/jstests/test-learning-resource.jsx [hunk 17]
```diff
@@ -119,76 +128,117 @@ define(['QUnit', 'jquery', 'react', 'lodash', 'learning_resources',
...
+            assert.equal($termsOptions.size(), 2);
+            assert.equal($termsOptions[0].selected, false);
+            assert.equal($termsOptions[1].selected, false);
+            // the second vocabulary can be a multi select
+            $($termsSelect)
```
Comment (noisecapella) [inline ui/jstests/test-learning-resource.jsx]: And `$()` here

File: ui/jstests/test-learning-resource.jsx (hunk 16 above)
Comment (jamiefolsom) [inline ui/jstests/test-learning-resource.jsx]: Did so, ran tests, and got: `TypeError: 'undefined' is not a function (evaluating '$termsSelect.find('option')')`

File: ui/jstests/test-learning-resource.jsx [hunk 18]
```diff
@@ -119,76 +128,117 @@ define(['QUnit', 'jquery', 'react', 'lodash', 'learning_resources',
...
+
+          assert.equal($vocabOptions[0].selected, true);
+          assert.equal($vocabOptions[1].selected, false);
 
           // TestUtils.Simulate.change only simulates a change event,
           // we need to update the value first ourselves
-          $($vocabSelect[0]).val("hard").trigger('change');
-          React.addons.TestUtils.Simulate.change($vocabSelect[0]);
+          $vocabSelect.val("prerequisite").trigger('change');
+          React.addons.TestUtils.Simulate.change($vocabSelect);
```
Comment (noisecapella) [inline ui/jstests/test-learning-resource.jsx:223]: `Simulate.change` will not work on a jquery object (see also `$termsSelect` below)

File: ui/jstests/test-learning-resource.jsx (hunk 16 above)
Comment (noisecapella) [inline ui/jstests/test-learning-resource.jsx]: Ok, it looks like you should change `$termsSelect` to just `termsSelect` instead and leave the `$()` on

Comment (jamiefolsom) [inline ui/jstests/test-learning-resource.jsx]: There's no var called termSelect:

```
ui/jstests/test-learning-resource.jsx: line 229, col 35, 'termsSelect' is not defined.
ui/jstests/test-learning-resource.jsx: line 234, col 15, 'termsSelect' is not defined.
```

Comment (noisecapella) [inline ui/jstests/test-learning-resource.jsx]: I mean rename `$termsSelect` to `termSelect` since it isn't a jquery object.

**Comment by noisecapella** [comment]
Code coverage looks good, functionality works excepting the case with zero vocabularies for a learning resource type. Clean implementation :)

File: ui/jstests/test-learning-resource.jsx (hunk 16 above)
Comment (jamiefolsom) [inline ui/jstests/test-learning-resource.jsx]: Ah, gotcha.

File: ui/static/ui/js/learning_resources.jsx (hunk 14 above)
Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx:99]: Done.

**Comment by jamiefolsom** [comment]
Great, I will handle the zero vocabulary case, then squash commits. Thanks, @noisecapella!

File: ui/static/ui/js/learning_resources.jsx (hunk 15 above)
Comment (jamiefolsom) [inline ui/static/ui/js/learning_resources.jsx]: This solves the proximate issue, avoiding accessing an attribute of an undefined object, but you raise a broader question, which we haven't dealt with, which is that the whole LR panel should not show this chooser UI if there are no vocabularies assigned to the resource type.

```
          if (vocabulariesAndTerms.length) {
            thiz.setState({
              selectedVocabulary: vocabulariesAndTerms[0].vocabulary
            });
          };
```

File: ui/jstests/test-learning-resource.jsx (hunk 18 above)
Comment (noisecapella) [inline ui/jstests/test-learning-resource.jsx:223]: It looks like you missed this here. The change event may be provided to React via `trigger()` in which case you can just remove the `Simulate.change` line.

Comment (noisecapella) [inline ui/jstests/test-learning-resource.jsx:223]: I'm fine merging anyway, we can do this in a separate PR

Comment (jamiefolsom) [inline ui/jstests/test-learning-resource.jsx:223]: OK, merging. Thanks!

**Comment by noisecapella** [comment]
Just commenting on `key` in React. It's used to simplify the tree diff operation by matching up old elements with new ones in their fake DOM. There was a presentation as OSCON that claimed a naive tree diff is O(n^3) but I can't remember why at the moment.
